skiko/src/androidMain/kotlin/org/jetbrains/skiko/Actuals.android.kt

```diff
@@ -16,7 +16,7 @@ internal class AndroidOpenGLRedrawer(
 ) : Redrawer {
     override fun dispose() = TODO()
     override fun needRedraw() = TODO()
-    override fun redrawImmediately() = TODO()
+    override fun redrawImmediately(waitForVsync: Boolean) = TODO()
```

Comment:
Without knowing your context, I am leaning towards not adding this parameter for configuration, and just change the default behaviour on macOs. I am not sure in which cases it is actually needed to wait for vsync in a case of `redrawImmediately`. 

Initially it was designed that this method shouldn't wait for vsync, and redraw the rendering buffer to the end of this method. Waiting for vsync on macOs was added because we could lose some frames if we don't wait.

Reply:
Yes, I started from that. The problem is that when we don't wait for VSync on window resize user experience slightly degrade. I mean that the gap between layer and window border become more noticable. So I would like to play with this settings a liitle bit

Reply:
>  redraw the rendering buffer to the end of this method

I.e., if vsync happens after this method, it is guarantied that we'll see the new content on the screen.

Reply:
>  when we don't wait for VSync on window resize

If we don't change the behaviour on resize (waiting for vsync), what is the case for `redrawImmediately(vsync = false)`?

In Compose we have 2 cases for `redrawImmediately`:
- window resize
- window showing (draw content before window is shown to avoid white/black blinking)

Also one more hypothetical case - manual calling to reduce input latency:
```
mouse move
change state
redrawImmediately

mouse move
change state
redrawImmediately

mouse move
change state
redrawImmediately

vsync
```

>  I mean that the gap between layer and window border become more noticable

This is strange. In my head waiting for something (vsync) should increase the gap, not decrease it.

Reply:
> This is strange. In my head waiting for something (vsync) should increase the gap

At my understanding blocking the paint method prevent window border from movement. Which means it may lag behind mouse cursor but move along with skia content (not perfectly).

> what is the case for redrawImmediately(vsync = false)?

I'm considering to change it on window resize in Fleet subclass of `SkiaLayer`. But I would like to have flexibility here, e.g. to be able to roll it back in case it would look too bad on 120hz displays. So it seems it may look different with different number of buffers in swap chain.

Reply:
It is okay to merge if it is more convenient to experiment in Fleet first.

But if it will be no big burden, and our goal is to fix glitches on resize, better to experiment with local Skiko (or some build from `a branch`). Because in the end we might not need this to be configurable. Also, the end solution should be applied to Compose as well, it means that resizing itself should be fixed in `SkiaLayer`, not in its subclass.

Configurable vsync here adds a burden of supporting it (people have to read/understand it) and implementing it properly (should `SkikoProperties.isVsyncEnabled` affect it, why is it different default values for different OS'es, what are use cases, etc)

Reply:
If we come to a conclusion that resize glitches are caused by vsync waiting, then we need to not only remove it, but also implement a proper triple buffering (see the other comment) that might help avoiding missed frames.

Reply:
> It is okay to merge if it is more convenient to experiment in Fleet first.

Let's merge it then. What you do in `SkiaLayer` will be the default behaviour. TBH that it will be a problem for other library users. 

> If we come to a conclusion that resize glitches are caused by vsync waiting

I think the main cause of resize issue is concurrency model of AWT (with two threads EDT and AppKitThread) with Java2D backed by OpenGL layer with Skia Metal layer on top of it.
In JWM straightforward approach works quite well https://github.com/HumbleUI/JWM/blob/5bdc87bca4ebbb25514794dcb866e801620ea71a/macos/cc/LayerMetal.mm#L76
https://github.com/user-attachments/assets/afbdfcd2-449e-4b13-8de3-abc6cf2d858f

I think that the right way to fix it would be to get rid of AWT. But for now I'm looking for a way to improve UX with what we have now.

Also I suspect that sometimes EDT in Fleet spend too much time on this locking, which lead to unresponsiveness. And maybe we could sacrifice some window resize ux if it turns to be true.

Reply:
> and implementing it properly 

Also, I noticed, that Windows/Linux has a different semantics - `wait for vsync after drawing`. On macOs we have `wait for vsync before drawing if the previous drawing was in this vsync`.

Do you need it for Windows/Linux? Can we just add a system property `skiko.rendering.macos.waitForPreviousFrameVsyncOnRedrawImmediately`?

> Let's merge it then. 

Before the merge we need to add a comment describing situation, to avoid future misunderstanding of this code:
```
/*
- wait for vsync before drawing if the previous drawing was in this vsync. Affects only `redrawImmediately` (called on window resize)
- affects only macOs
- this is for experiments for window resize
- setting it false may cause missing frames https://github.com/JetBrains/skiko/pull/708
*/
```
If we decide to implement Windows/Linux, I would implement it as another property:
```
Wait for vsync after drawing. Affects only `redrawImmediately` (called on window resize)

Windows/Linux also depends on SkikoProperties.isVsyncEnabled (greater priority)
```